clippy_lints/src/types.rs

```diff
@@ -1598,7 +1600,9 @@ impl<'tcx> LateLintPass<'tcx> for Casts {
         if let ExprKind::Cast(ref ex, _) = expr.kind {
             let (cast_from, cast_to) = (cx.typeck_results().expr_ty(ex), cx.typeck_results().expr_ty(expr));
             lint_fn_to_numeric_cast(cx, expr, ex, cast_from, cast_to);
-            if let ExprKind::Lit(ref lit) = ex.kind {
+            if let Some(lit) = get_numeric_literal(ex) {
```

Comment:
Since this affects the old behavior of the lint too, we should add the following cases to the fixable test:
```rust
    let _ = -100 as f32;
    let _ = -100 as f64;
    let _ = -100_i32 as f64;
```
Currently we lint those, but the suggestion loses the sign:
```
+error: casting integer literal to `f32` is unnecessary
+  --> $DIR/unnecessary_cast_fixable.rs:36:13
+   |
+LL |     let _ = -100 as f32;
+   |             ^^^^^^^^^^^ help: try: `100_f32`
+
```

Reply:
Done